pkg/oauth/apiserver/registry/personaltokens/delegate/delegate.go

```diff
@@ -0,0 +1,186 @@
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken LIST result to its typed version"))
+	}
+
+	return oauthAccessTokenListToUserOAuthAccessTokenList(tokenList), nil
+}
+
+func (r *REST) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenGenericObj, err := r.accessTokenStorage.Get(ctx, name, options)
+	if err != nil {
+		return nil, err
+	}
+
+	token := tokenGenericObj.(*oauthapi.OAuthAccessToken)
+	if token == nil {
+		return nil, errors.NewInternalError(fmt.Errorf("failed to convert generic accesstoken GET result to its typed version"))
+	}
+
+	if token.UserName != ctxUserName { // don't reveal other people tokens' hashes
+		return nil, errors.NewNotFound(oauth.Resource("personaltoken"), name)
+	}
+
+	return &oauthapi.PersonalAccessToken{OAuthAccessToken: *token}, nil
+}
+
+func (r *REST) Watch(ctx context.Context, options *metainternal.ListOptions) (watch.Interface, error) {
+	if ctx == nil {
+		return nil, fmt.Errorf("context is nil")
+	}
+
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sanitizedListOpts, err := userFieldSelectorToListOptions(options, ctxUserName)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenListWatcher, err := r.accessTokenStorage.Watch(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher := newOAuthAccessTokenWatcher(tokenListWatcher.ResultChan())
+	go watcher.Watch()
+	return watcher, nil
+}
+
+func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	return r.tableConvertor.ConvertToTable(ctx, object, tableOptions)
+}
+
+func (r *REST) NamespaceScoped() bool {
+	return false
+}
+
+func (r *REST) Delete(ctx context.Context, name string, validateFunc rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	if _, err := r.Get(ctx, name, &metav1.GetOptions{}); err != nil {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, err
+	}
+
+	return r.accessTokenStorage.Delete(ctx, name, validateFunc, options)
+}
+
+func getUserFromContext(ctx context.Context) (string, error) {
+	userInfo, exists := apirequest.UserFrom(ctx)
+	if !exists {
+		return "", kerrors.NewForbidden(oauth.Resource("personaltokens"), "", fmt.Errorf("unable to get personal tokens without a user in context"))
+	}
+
+	userName := userInfo.GetName()
+	if len(userName) == 0 {
+		return "", kerrors.NewForbidden(oauth.Resource("personaltokens"), "", fmt.Errorf("user has a zero-length username"))
+	}
+
+	return userName, nil
+}
+
+func oauthAccessTokenListToUserOAuthAccessTokenList(l *oauthapi.OAuthAccessTokenList) *oauthapi.PersonalAccessTokenList {
+	ret := oauthapi.PersonalAccessTokenList{}
+	for _, t := range l.Items {
+		ret.Items = append(ret.Items, oauthapi.PersonalAccessToken{OAuthAccessToken: t})
```

Comment:
do we use this pattern of embedding elsewhere? If not, I would be slightly concerned that things might break if TypeMeta and ObjectMeta are not at the toplevel of the type.

Reply:
I assume it's novel, I wasn't looking for it elsewhere

Reply:
we will see. If you have it working, maybe everything is fine.